src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,16 +1899,26 @@ public sealed class One : Tests
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        lengthTest: out Tests lengthTest,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    whenTrue = trueDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : trueDecisionPermitsTrueOther ? this : False.Instance;
+                    whenFalse = falseDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : falseDecisionPermitsTrueOther ? this : False.Instance;
```

Comment:
The problem was that we had completely replaced the original test.. this is now truly conditional on the length value.

If the current test subsumes the other (e.g. either of these is true), we insert this sequence so that when `L=N` we won't check anything else further, otherwise we emit the original test as usual.
e.g. `[_, >0, ..] or [.., <=0, _]` becomes `length >= 2 && [1] > 0 && (length == 3 || [^1] <= 0)`

Note that this test could get subsumed entirely if the we have a fixed list further down the line
e.g. `[.., 42] or [42]` becomes `length >=1 && [^1] is 42`

If there's no remaining value left for this element, the length alone gets emitted.
e.g. `[.., >0] or [<=0]` becomes `length >=1 && ([^1] is >0 || length == 1)`

In the above example the length test has been subsumed with the one from the pattern, but that's not always the case.
e.g. `[.., >0] or [<=0, ..]` becomes `length >= 1 && ([^1] > 0 || length == 1 || [0] <= 0)`

If there's multiple tests for each element, we do this for all of them, the pattern-matching machinery can merge any duplicated tests inserted into the tree.

There may be some optimization opportunities to omit redundant indexer accesses which can be done separately.

Compared to the other approaches we discussed, I think this is the most efficient in both implementation and codegen, because we only do this when we know it's absolutely needed with no preprocessing or reordering of tests.